### PR TITLE
Establish a best practice for trans dependency management.

### PR DESCRIPTION
**Pull request by jmchilton**
As part of the ongoing effort to establish a best practice for modern backend components - I think this is important step. I could explain it all but I think the docstring at the top of the PR does a pretty good job explaining why this is a good direction.

### CODE REVIEW

File: lib/galaxy/managers/context.py [hunk 1]
```diff
@@ -1,6 +1,39 @@
...
+
+These objects describe the context around a unit of work. This unit of work
+is very broad and can be anything from the response to a web request, the
+scheduling of a workflow, the reloading the toolbox, etc.. Traditionally,
+Galaxy has simply passed around a GalaxyWebTransaction object through
+all layers and large components of the Galaxy app. Having random backend
+components define explicit dependencies on this however is inappropriate
+because Galaxy may not be used in all sort of non-web contexts. The future
+use of message queues and web sockets as well as the decomposition of the
+backend into packages all only further make this heavy reliance on
```
Comment (nsoranzo) [inline lib/galaxy/managers/context.py]:
```suggestion
backend into packages only further make this heavy reliance on
```

File: lib/galaxy/managers/context.py [hunk 2]
```diff
@@ -99,14 +140,18 @@ def install_model(self):
...
+    """ For transaction-like objects to provide Galaxy convenience layer for
     reasoning about users.
 
     Mixed in class must provide `user` and `app`
     properties.
     """
 
+    @abc.abstractproperty
+    def user(self):
+        """Provide access to the app object."""
```
Comment (nsoranzo) [inline lib/galaxy/managers/context.py]:
```suggestion
        """Provide access to the user object."""
```

**Comment by nsoranzo** [comment]
The py36-unit build failure seems relevant.

File: lib/galaxy/webapps/galaxy/api/roles.py [hunk 3]
```diff
@@ -65,26 +63,26 @@ class FastAPIRoles:
     role_manager: RoleManager = Depends(get_role_manager)
 
     @router.get('/api/roles')
-    def index(self, trans: SessionRequestContext = Depends(get_trans)) -> RoleListModel:
+    def index(self, trans: ProvidesUserContext = Depends(get_trans)) -> RoleListModel:
```
Comment (jdavcs) [inline lib/galaxy/webapps/galaxy/api/roles.py:66]: Maybe move the `RoleListModel` definition (line 43) to `galaxy/manager/roles` and import it alongside the other pydantic models?

Comment (jmchilton) [reply lib/galaxy/webapps/galaxy/api/roles.py:66]: I get why that seems more wholistic and consistent. My counter argument is that I think the distinction is that RoleListModel exists only to adapt a list of roles to the web framework - the managers themselves are more useful and better when they return a list of role models. So it feels like moving that there is pushing "web app concerns" into "app" logic.

If you don't buy that though, I'll definitely refactor this to and push that down into the managers.

Comment (jdavcs) [reply lib/galaxy/webapps/galaxy/api/roles.py:66]: No, I agree, you're right. In fast, I had similar misgivings about the `TourList` model being part of the tours "schema". I think once we have a good sample of endpoints done, we may go back and refactor for consistency.

**Review by jdavcs — APPROVED** [review]

File: lib/galaxy/webapps/galaxy/api/roles.py (hunk 3 above)
Comment (jmchilton) [reply lib/galaxy/webapps/galaxy/api/roles.py:66]: Finding the right distinction here is messy - I think we're asking the right questions and things are structured better than ever before inside the app - but we will need to work on this and accept a lot of fuzziness in these distinctions.

Thanks for the detailed review!

**Pull request merged**